node/full_node_test.go

```diff
@@ -25,78 +25,102 @@ import (
-	defer cancel()
-	node, err := newFullNode(ctx, config.NodeConfig{DALayer: "mock"}, key, signingKey, proxy.NewLocalClientCreator(app), &types.GenesisDoc{ChainID: "test"}, log.TestingLogger())
-	require.NoError(err)
-	require.NotNil(node)
-
-	assert.False(node.IsRunning())
-
-	err = node.Start()
-	assert.NoError(err)
-	defer func() {
-		assert.NoError(node.Stop())
-	}()
+	ctx := context.Background()
+	node := initializeAndStartNode(ctx, require, assert)
+	defer cleanUpNode(node, assert)
 	assert.True(node.IsRunning())
 }
 
 func TestMempoolDirectly(t *testing.T) {
 	assert := assert.New(t)
 	require := require.New(t)
 
+	ctx := context.Background()
+	node := initializeAndStartNode(ctx, require, assert)
+	defer cleanUpNode(node, assert)
+
+	peerID := getPeerID(assert)
+	verifyTransactions(node, peerID, assert)
+	verifyMempoolSize(node, assert)
+}
+
+// cleanUpNode stops the node and checks if it is running
+func cleanUpNode(node *FullNode, assert *assert.Assertions) {
+	defer node.cancel()
+	assert.NoError(node.Stop())
+	assert.False(node.IsRunning())
+}
+
+// setupMockApplication initializes a mock application
+func setupMockApplication() *mocks.Application {
 	app := &mocks.Application{}
 	app.On(InitChain, mock.Anything).Return(abci.ResponseInitChain{})
 	app.On(CheckTx, mock.Anything).Return(abci.ResponseCheckTx{})
+	return app
+}
+
+// generateSingleKey generates a single private key
+func generateSingleKey() crypto.PrivKey {
 	key, _, _ := crypto.GenerateEd25519Key(rand.Reader)
-	signingKey, _, _ := crypto.GenerateEd25519Key(rand.Reader)
-	anotherKey, _, _ := crypto.GenerateEd25519Key(rand.Reader)
+	return key
+}
+
+// newTestNode creates a new test node
+func newTestNode(ctx context.Context, key crypto.PrivKey, signingKey crypto.PrivKey, app *mocks.Application) (*FullNode, error) {
+	return newFullNode(ctx, config.NodeConfig{DALayer: "mock"}, key, signingKey, proxy.NewLocalClientCreator(app), &types.GenesisDoc{ChainID: "test"}, log.TestingLogger())
+}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	node, err := newFullNode(ctx, config.NodeConfig{DALayer: "mock"}, key, signingKey, proxy.NewLocalClientCreator(app), &types.GenesisDoc{ChainID: "test"}, log.TestingLogger())
+// setupTestNode sets up a test node
+func setupTestNode(ctx context.Context, require *require.Assertions) *FullNode {
+	app := setupMockApplication()
+	key, signingKey := generateSingleKey(), generateSingleKey()
+	node, err := newTestNode(ctx, key, signingKey, app)
 	require.NoError(err)
 	require.NotNil(node)
+	return node
+}
 
-	err = node.Start()
-	require.NoError(err)
-	defer func() {
-		assert.NoError(node.Stop())
-	}()
+// initializeAndStartNode initializes and starts a test node
+func initializeAndStartNode(ctx context.Context, require *require.Assertions, assert *assert.Assertions) *FullNode {
+	node := setupTestNode(ctx, require)
+	assert.False(node.IsRunning())
+	err := node.Start()
+	assert.NoError(err)
+	return node
```

Comment:
might as well put the `assert.True(node.IsRunning())` here. Then the core functionality is always verified in every test.  Plus it verifies that the test helper is functioning properly.  

Then `TestStartup` doesn't need the check.

Reply:
Great catch, thanks!